gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixUtils.java

```diff
@@ -89,34 +90,43 @@ public static String getHelixInstanceName(
     return namePrefix + "_" + instanceId;
   }
 
+  // We have switched from Helix JobQueue to WorkFlow based job execution.
+  @Deprecated
   public static void submitJobToQueue(
       JobConfig.Builder jobConfigBuilder,
       String queueName,
       String jobName,
       TaskDriver helixTaskDriver,
       HelixManager helixManager,
       long jobQueueDeleteTimeoutSeconds) throws Exception {
+    submitJobToWorkFlow(jobConfigBuilder, queueName, jobName, helixTaskDriver, helixManager, jobQueueDeleteTimeoutSeconds);
+  }
 
+  public static void submitJobToWorkFlow(JobConfig.Builder jobConfigBuilder,
+      String queueName,
+      String jobName,
+      TaskDriver helixTaskDriver,
+      HelixManager helixManager,
+      long jobQueueDeleteTimeoutSeconds) throws Exception {
     WorkflowConfig workflowConfig = helixTaskDriver.getWorkflowConfig(helixManager, queueName);
-
     // If the queue is present, but in delete state then wait for cleanup before recreating the queue
     if (workflowConfig != null && workflowConfig.getTargetState() == TargetState.DELETE) {
       new TaskDriver(helixManager).deleteAndWaitForCompletion(queueName, jobQueueDeleteTimeoutSeconds);
       // if we get here then the workflow was successfully deleted
       workflowConfig = null;
     }
 
-    // Create one queue for each job with the job name being the queue name
+    // Create a work flow for each job with the name being the queue name
     if (workflowConfig == null) {
-      JobQueue jobQueue = new JobQueue.Builder(queueName).build();
-      helixTaskDriver.createQueue(jobQueue);
-      log.info("Created job queue {}", queueName);
+      // Create a workflow, set expiry to 5 minutes and add the job
+      workflowConfig = new WorkflowConfig.Builder().setExpiry(5L, TimeUnit.MINUTES).build();
```

Comment:
Can we move this expiration to next PR so that this PR only contains the migration from jobQueue to workflow?

If I remember correctly, Helix team want us to use the stop method and then fail the job on purpose via some specific API so that the delete will be handled by Helix properly.

Reply:
Even if we fail the job with some new Helix API, there will always be race condition between job stop->fail->cleanup and job status checking in HelixUtils::waitJobCompletion().
if waitJobCompletion could not check status between fail and cleanup, it will stuck there; and this is very likely.
I think we should go with adding a synchronization primitive as you suggested offline.

Reply:
Yes, let's handle this in our next PR: 
   1) Auto deletion
      a) Add expiration attributes.
      b) stop->fail->cleanup
   2) An additional status flag to avoid race condition
Once we have auto deletion ready, we can
   3) Remove workflow deletion in the job submission (but you have to use jobId as the workflow name so a non-deleted workflow won't block the new jobs.)